libandroid-navigation-ui/src/main/java/com/mapbox/services/android/navigation/ui/v5/NavigationViewModel.java

```diff
@@ -97,15 +99,19 @@ public NavigationViewModel(Application application) {
     this.connectivityController = new MapConnectivityController();
   }
 
+  @TestOnly
```

Comment:
should we be concerned about the fact that the "test" constructor handles object initialization differently than public one?

Reply:
I don't think so. I tried to test the public constructor, but it needs a MapView to be drawn first before it can even start testing the constructor. Also, we need to make sure is that we have mocked these objects so that we can test the state of objects depending on their initialization. `@TestOnly` would make sure that no 3rd party SDK users can misuse them.

Reply:
i'm curious about thought process here: if the init exec path for the object is different during in test and prod how do we ensure prod is not regressed as a result of this change? also is it only a matter of adding mapview? they really do look differently.

Reply:
Yes, I agree the init exec path is different. I am sure we only need a MapView, because the line `Mapbox.getAccessToken()` internally needs a MapView. The unit test for public constructor fails at this line because of the absence of `MapView`. If you look at the initialization of these objects, the `NavigationViewModel` initializes a lot of these objects from this method https://github.com/mapbox/mapbox-navigation-android/blob/d9789788bba8432b02ceefa18ff5c62ed328fe1c/libandroid-navigation-ui/src/main/java/com/mapbox/services/android/navigation/ui/v5/NavigationViewModel.java#L209. This method is called from `NavigationView/startNavigation`. A lot of our unit test doesn't depend on whether the navigation has started. The test constructors just makes sure that we have these objects beforehand rather than initializing them only when you start navigation.

I myself don't like the idea of spinning up the Test constructors, but to be able to test the public constructor it will have to be a mix of instrumentation+unit testing. Also, then for mostly all the unit test we would first have to start the navigation, because only then we will have all the objects initialized. I feel that's the right way to do, but it is a very big change. 

I added the test here to make sure that the piece of code I had written is getting properly covered. But, we need to fix this with another ticket(probably in 1.0 UI SDK) where the views will be tested. Since `Mapbox.getAccessToken()` internally references to `MapView` and needs a `Context` probably `NavigationViewModel` is not a good place to put it in, since we don't have access to the context here. It is a much bigger refactor

Reply:
Yeah as @abhishek1508 outlines nicely here, the delayed initialization (waiting for `NavigationViewOptions`) is not ideal.

This may be a bigger refactor / out of scope for this PR, but a `ViewModelFactory` may help solve a lot of the problems here. 

```
class NavigationViewModelFactory(val accessToken: String, ...) : ViewModelProvider.Factory {
    override fun <T : ViewModel?> create(modelClass: Class<T>): T {
        return NavigationViewModel(accessToken) as T
    }
}
```

The benefit here is not only that we can test against one constructor, but also we can extract a lot of the initialization code from the `NavigationViewModel` itself and wait for the `NavigationViewOptions` before we try to retrieve it at all. 

```
NavigationViewOptions options = ...
// TODO All things that need to be initialized for view model here

NavigationViewModelFactory factory = new NavigationViewModelFactory(Mapbox.getAccessToken(), ...);

navigationViewModel = ViewModelProviders.of(this, factory).get(NavigationViewModel.class);
```

The problem here is then everywhere in `NavigationView` where we assume `navigationViewModel` is not null (because we currently retrieve it right away [here](https://github.com/mapbox/mapbox-navigation-android/blob/71fae584248a7b9c703fe9232bdd70fa6df7dee2/libandroid-navigation-ui/src/main/java/com/mapbox/services/android/navigation/ui/v5/NavigationView.java#L499)).  That would also have to be refactored.